conda/match.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from .models.dist import Dist
+from .models.index_record import IndexRecord
+from .common.compat import string_types, text_type, iteritems
+from .version import VersionSpec
+from .exceptions import CondaValueError
+from ast import literal_eval
+import re
+import sys
+
+class MatchSpec(object):
+
+    def __new__(cls, *specs, **kwargs):
+        if len(specs) == 1 and not kwargs and isinstance(specs[0], cls):
+            return specs[0]
+        normalize = kwargs.pop('normalize', False)
+        self = object.__new__(cls)
+        self.specs = {}
+
+        def push_(*args):
```

Comment:
Add comment about each arg being a key, value pair.  Somewhere nearby make (or reference) a complete list of valid fields.